spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/FrameManager.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.Graph.ModuleNode
+import com.intel.analytics.bigdl.nn.ops.{Exit, MergeOps, NextIteration}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+/**
+ * Manage frame in scheduler. When scheduler execute nodes, it may enter a `frame`. Before
+ * scheduler leave a frame, it must make sure all nodes in that frames has been run.
+ * @tparam T
+ */
+class FrameManager[T] extends Serializable {
+  import FrameManager._
+
+  /**
+   * Create a frame. If it exists, nothing happen
+   *
+   * @param frame
+   * @param parentFrame
+   */
+  def createFrame(frame: String, parentFrame: Option[Frame[T]]): Frame[T] = {
+    if (!frames.contains(frame)) {
+      frames(frame) = new Frame(frame, parentFrame)
+    }
+    frames(frame)
+  }
+  /**
+   * Mark a node enter into frame. The node cannot be in two frames at the same time
+   * @param node node
+   * @param frame the name of the frame
+   */
+  def enter(node: ModuleNode[T], frame : Frame[T]): Unit = {
+    val name = node.element.getName()
+    if (nodeFrame.contains(name)) {
+      require(nodeFrame(name).eq(frame), "node cannot be in two different fames at the same time")
+    } else {
+      nodeFrame(name) = frame
+    }
+
+    if (!frame.nodes.contains(node)) {
+      if (addToFrameNode(node, frame)) frame.nodes.append(node)
+    }
+  }
+
+  def pend(node: ModuleNode[T], frame : Frame[T]): Unit = {
+    val name = node.element.getName()
+    require(node.element.isInstanceOf[NextIteration[_, _]], "you can only pend next iteration node")
+    if (nodeFrame.contains(name)) {
+      require(nodeFrame(name).eq(frame), "node cannot be in two different fames at the same time")
+    } else {
+      nodeFrame(name) = frame
+    }
+
+    frame.barrier -= 1
+    frame.waitingNodes.append(node)
+  }
+
+  /**
+   * Add nodes which will be executed many times in the loop belong to the given frame
+   * @param node
+   * @param frame
+   * @return
+   */
+  private def addToFrameNode(node: ModuleNode[T], frame : Frame[T]): Boolean = {
```

Comment:
It seems the method name and document does not match the actual behavior of this method. It just return a predicate but does not do the "addToFrame", right?

Reply:
change doc and rename the method to "isExecuteManyTimes"